### PR TITLE
rpc: subscription filters validity check

### PR DESCRIPTION
**Pull request by AliceInHunterland**
Change BlockFilter Primary field from int to byte. Additional check of filters parameters added for filter validation.

Closes: #3241

### CODE REVIEW

File: docs/notifications.md [hunk 1]
```diff
@@ -64,7 +64,7 @@ omitted if empty).
 
 Recognized stream names:
  * `block_added`
-   Filter: `primary` as an integer with primary (speaker) node index from
+   Filter: `primary` as an byte with primary (speaker) node index from
```
Comment (AnnaShaleva) [inline docs/notifications.md:67]: Adjust the commit title: we used to see an action in the title, e.g. fix something or implement something or add some functionality. So it's probably better be: `rpc: change the type of BlockFilter's Primary field to byte` or `rpc: change the type of BlockFilter.Primary from int to byte`

Also there's a typo in the commit description: s/`It needs to`/`It's needed to` and s/`to prevent change filter's`/`to prevent changing filters`?

File: pkg/neorpc/filters.go [hunk 2]
```diff
@@ -48,6 +51,43 @@ type (
 		Type   *mempoolevent.Type `json:"type,omitempty"`
 	}
 )
+type ValidFilter interface {
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:54]: Missing new line.

Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:54]: s/`ValidFilter`/`SubscriptionFilter`

Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:54]: Exported structure should have a comment.

File: pkg/neorpc/filters.go [hunk 3]
```diff
@@ -48,6 +51,43 @@ type (
...
+	IsValid() error
+}
+
+// IsValid checks whether the filter is valid.
+func (f *BlockFilter) IsValid() error {
+	return nil
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:60]: You can move this comment to `NotificationFilter` interface, and here add just a `IsValid implements NotificationFilter interface.` note. Usually it's done in order not to change every single implementation documentation in case if something is changed in the logic description.

File: pkg/neorpc/filters.go [hunk 4]
```diff
@@ -48,6 +51,43 @@ type (
 		Type   *mempoolevent.Type `json:"type,omitempty"`
 	}
 )
+type ValidFilter interface {
+	IsValid() error
+}
+
+// IsValid checks whether the filter is valid.
+func (f *BlockFilter) IsValid() error {
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:59]: Move it under `(f *BlockFilter) Copy()`, and move the rest of implementations correspondingly.

File: pkg/neorpc/filters.go [hunk 5]
```diff
@@ -48,6 +51,43 @@ type (
...
+}
+
+// IsValid checks whether the filter is valid.
+func (f *TxFilter) IsValid() error {
+	return nil
+}
+
+// IsValid checks whether the filter is valid.
+func (f *NotificationFilter) IsValid() error {
+	if f.Name != nil && runtime.MaxEventNameLen < len(*f.Name) {
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:70]: s/`runtime.MaxEventNameLen < len(*f.Name)`/`len(*f.Name) > runtime.MaxEventNameLen`? It's just more usual way.

File: pkg/neorpc/filters.go [hunk 6]
```diff
@@ -48,6 +51,43 @@ type (
...
+		return errors.New("bad name parameter")
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:71]: return fmt.Errorf("NotificationFilter name parameter must be less than %d", runtime.MaxEventNameLen)

File: pkg/neorpc/filters.go [hunk 7]
```diff
@@ -48,6 +51,43 @@ type (
...
+	}
+	return nil
+}
+
+// IsValid checks whether the filter is valid.
+func (f *ExecutionFilter) IsValid() error {
+	if f.State != nil {
+		if *f.State != "HALT" && *f.State != "FAULT" {
+			return errors.New("bad state parameter")
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:80]: return fmt.Errorf("ExecutionFilter state parameter must be either %s or %s", vmstate.Halt, vmstate.Fault)

File: pkg/neorpc/filters.go [hunk 8]
```diff
@@ -48,6 +51,43 @@ type (
...
+	if f.Name != nil && runtime.MaxEventNameLen < len(*f.Name) {
+		return errors.New("bad name parameter")
+	}
+	return nil
+}
+
+// IsValid checks whether the filter is valid.
+func (f *ExecutionFilter) IsValid() error {
+	if f.State != nil {
+		if *f.State != "HALT" && *f.State != "FAULT" {
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:79]: Replace with vmstate.Halt.String() and vmstate.Fault.String() resp.?

File: pkg/rpcclient/wsclient.go [hunk 9]
```diff
@@ -795,11 +795,6 @@ func (c *WSClient) ReceiveExecutions(flt *neorpc.ExecutionFilter, rcvr chan<- *s
 	}
 	params := []any{"transaction_executed"}
 	if flt != nil {
-		if flt.State != nil {
-			if *flt.State != "HALT" && *flt.State != "FAULT" {
-				return "", errors.New("bad state parameter")
-			}
-		}
```
Comment (AnnaShaleva) [inline pkg/rpcclient/wsclient.go:802]: So instead of this removed check you have to add client-side IsValid check. Add this check to `(c *WSClient) performSubscription`. Change `Comparator` interface so that `Filter() any` returns NotificationFilter interface.

File: pkg/services/rpcsrv/server.go [hunk 10]
```diff
@@ -2729,29 +2729,24 @@ func (s *Server) subscribe(reqParams params.Params, sub *subscriber) (any, *neor
 		param := *p
 		jd := json.NewDecoder(bytes.NewReader(param.RawMessage))
 		jd.DisallowUnknownFields()
+		flt := neorpc.NewFilter(event)
+		err = flt.IsValid()
+		if err != nil {
+			return nil, neorpc.WrapErrorWithData(neorpc.ErrInvalidParams, err.Error())
+		}
+		err = jd.Decode(flt)
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/server.go:2737]: You don't need this change, just leave this code as is and replace `var filter any` with `var filter NotificationFilter`. After decoding call `filter.IsValid`.

File: pkg/neorpc/filters.go (hunk 2 above)
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:54]: Change the commit title to something like `neorpc: add WS notification filter IsValid functionality`, otherwise it's not that clear what's the `filter` in this scope.

File: pkg/neorpc/filters.go [hunk 11]
```diff
@@ -49,6 +53,12 @@ type (
 	}
 )
 
+// SubscriptionFilter is an interface for all subscription filters.
+// IsValid checks whether the filter is valid.
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:57]: `IsValid` comment may be placed right above `IsValid()` method.

Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:57]: IsValid checks whether the filter is valid and returns a specific [ErrInvalidSubscriptionFilter] error if not.

File: pkg/neorpc/filters.go [hunk 12]
```diff
@@ -49,6 +53,12 @@ type (
...
+type SubscriptionFilter interface {
+	IsValid() error
+}
+
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:61]: Create `var ErrInvalidSubscriptionFilter = errors.New("invalid subscription filter")` here with appropriate comment and return it as a wrapped error everywhere.

File: pkg/neorpc/filters.go [hunk 13]
```diff
@@ -104,6 +124,14 @@ func (f *NotificationFilter) Copy() *NotificationFilter {
 	return res
 }
 
+// IsValid implements SubscriptionFilter interface.
+func (f *NotificationFilter) IsValid() error {
+	if f.Name != nil && len(*f.Name) > runtime.MaxEventNameLen {
+		return fmt.Errorf("NotificationFilter name parameter must be less than %d", runtime.MaxEventNameLen)
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:130]: I.e. here wrapped error will look like: `fmt.Errorf("%w: NotificationFilter name ...", ErrInvalidSubscriptionFilter, ...)`

File: pkg/neorpc/rpcevent/filter.go [hunk 14]
```diff
@@ -35,14 +35,22 @@ func Matches(f Comparator, r Container) bool {
 	}
 	switch f.EventID() {
 	case neorpc.BlockEventID:
-		filt := filter.(neorpc.BlockFilter)
+		err := filter.IsValid()
```
Comment (AnnaShaleva) [inline pkg/neorpc/rpcevent/filter.go:38]: You don't need to include this check into `Matches`. Instead, add this check to the client-side and to the server-side.

Client-side code should be exactly like in https://github.com/nspcc-dev/neo-go/pull/3258/commits/3840e506719223cb2f7ecc356d2e1446cf2fa908#r1425034760 described:
```
func (c *WSClient) performSubscription(params []any, rcvr notificationReceiver) (string, error) {
	var resp string

	if flt := rcvr.Filter(); flt != nil {
		err := flt.IsValid()
		if err != nil {
			return "", err
		}
	}
```

And the server-side code is OK now.

File: pkg/neorpc/filters.go [hunk 15]
```diff
@@ -70,6 +80,11 @@ func (f *BlockFilter) Copy() *BlockFilter {
 	return res
 }
 
+// IsValid implements SubscriptionFilter interface.
+func (f *BlockFilter) IsValid() error {
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:84]: Here and below define it on struct, not on pointer.

File: pkg/neorpc/rpcevent/filter.go [hunk 16]
```diff
@@ -35,14 +35,22 @@ func Matches(f Comparator, r Container) bool {
...
+		if err != nil {
+			return false
+		}
+		filt := filter.(*neorpc.BlockFilter)
```
Comment (AnnaShaleva) [inline pkg/neorpc/rpcevent/filter.go:42]: Useless change, keep the `filt := filter.(neorpc.BlockFilter)`.

File: pkg/neorpc/rpcevent/filter_test.go [hunk 17]
```diff
@@ -117,7 +117,7 @@ func TestMatches(t *testing.T) {
 			name: "block, primary mismatch",
 			comparator: testComparator{
 				id:     neorpc.BlockEventID,
-				filter: neorpc.BlockFilter{Primary: &badPrimary},
+				filter: neorpc.SubscriptionFilter(&neorpc.BlockFilter{Primary: &badPrimary}),
```
Comment (AnnaShaleva) [inline pkg/neorpc/rpcevent/filter_test.go:120]: Useless change. If IsValid is defined not on the pointer, then it should be OK without `neorpc.SubscriptionFilter` assertion.

File: pkg/rpcclient/wsclient.go [hunk 18]
```diff
@@ -127,11 +127,11 @@ func (r *blockReceiver) EventID() neorpc.EventID {
 }
 
 // Filter implements neorpc.Comparator interface.
-func (r *blockReceiver) Filter() any {
+func (r *blockReceiver) Filter() neorpc.SubscriptionFilter {
 	if r.filter == nil {
 		return nil
 	}
-	return *r.filter
+	return r.filter
```
Comment (AnnaShaleva) [inline pkg/rpcclient/wsclient.go:134]: Revert it here and below.

File: pkg/rpcclient/wsclient.go [hunk 19]
```diff
@@ -711,7 +711,12 @@ func (c *WSClient) makeWsRequest(r *neorpc.Request) (*neorpc.Response, error) {
 
 func (c *WSClient) performSubscription(params []any, rcvr notificationReceiver) (string, error) {
 	var resp string
-
```
Comment (AnnaShaleva) [inline pkg/rpcclient/wsclient.go:714]: Useless change.

File: pkg/rpcclient/wsclient.go [hunk 20]
```diff
@@ -711,7 +711,12 @@ func (c *WSClient) makeWsRequest(r *neorpc.Request) (*neorpc.Response, error) {
...
+	if rcvr.Filter() != nil {
+		err := rcvr.Filter().IsValid()
+		if err != nil {
+			return "", err
+		}
+	}
```
Comment (AnnaShaleva) [inline pkg/rpcclient/wsclient.go:719]: Good, I see it's here. You may simplify this check by a single `if flt := rcvr.Filter(); flt != nil {` and then use `flt` inside `if`.

File: pkg/services/rpcsrv/server.go [hunk 21]
```diff
@@ -2733,24 +2733,24 @@ func (s *Server) subscribe(reqParams params.Params, sub *subscriber) (any, *neor
 		case neorpc.BlockEventID:
 			flt := new(neorpc.BlockFilter)
 			err = jd.Decode(flt)
-			filter = *flt
+			filter = flt
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/server.go:2736]: Revert this change.

File: pkg/services/rpcsrv/server.go [hunk 22]
```diff
@@ -2759,7 +2759,12 @@ func (s *Server) subscribe(reqParams params.Params, sub *subscriber) (any, *neor
 			return nil, neorpc.WrapErrorWithData(neorpc.ErrInvalidParams, err.Error())
 		}
 	}
-
+	if filter != nil {
+		err = filter.IsValid()
+		if err != nil {
+			return nil, neorpc.WrapErrorWithData(neorpc.ErrInvalidParams, err.Error())
+		}
+	}
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/server.go:2767]: Empty line after this block.

**Review by AnnaShaleva — COMMENTED** [review]
Please, remove "TODO" from the PR description.

File: pkg/neorpc/filters.go [hunk 23]
```diff
@@ -104,6 +127,14 @@ func (f *NotificationFilter) Copy() *NotificationFilter {
 	return res
 }
 
+// IsValid implements SubscriptionFilter interface.
+func (f NotificationFilter) IsValid() error {
+	if f.Name != nil && len(*f.Name) > runtime.MaxEventNameLen {
+		return fmt.Errorf("NotificationFilter name parameter must be less than %d. %w", runtime.MaxEventNameLen, ErrInvalidSubscriptionFilter)
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:133]: See the https://github.com/nspcc-dev/neo-go/pull/3258#discussion_r1426347845. Firstly `%w`, after then additional error test.

File: pkg/neorpc/filters.go [hunk 24]
```diff
@@ -121,6 +152,17 @@ func (f *ExecutionFilter) Copy() *ExecutionFilter {
 	return res
 }
 
+// IsValid implements SubscriptionFilter interface.
+func (f ExecutionFilter) IsValid() error {
+	if f.State != nil {
+		if *f.State != vmstate.Halt.String() && *f.State != vmstate.Fault.String() {
+			return fmt.Errorf("ExecutionFilter state parameter must be either %s or %s. %w", vmstate.Halt, vmstate.Fault, ErrInvalidSubscriptionFilter)
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:159]: Ditto.

File: pkg/services/rpcsrv/server.go [hunk 25]
```diff
@@ -2759,6 +2759,12 @@ func (s *Server) subscribe(reqParams params.Params, sub *subscriber) (any, *neor
 			return nil, neorpc.WrapErrorWithData(neorpc.ErrInvalidParams, err.Error())
 		}
 	}
+	if filter != nil {
+		err = filter.IsValid()
+		if err != nil {
+			return nil, neorpc.WrapErrorWithData(neorpc.ErrInvalidParams, err.Error())
+		}
+	}
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/server.go:2767]: Good, we need to test this server-side functionality.

File: pkg/rpcclient/wsclient.go [hunk 26]
```diff
@@ -712,6 +712,11 @@ func (c *WSClient) makeWsRequest(r *neorpc.Request) (*neorpc.Response, error) {
 func (c *WSClient) performSubscription(params []any, rcvr notificationReceiver) (string, error) {
 	var resp string
 
+	if flt := rcvr.Filter(); flt != nil {
+		if err := flt.IsValid(); err != nil {
+			return "", err
+		}
+	}
```
Comment (AnnaShaleva) [inline pkg/rpcclient/wsclient.go:719]: And the client-side functionality should also be tested. There's already test for VM states, need to add a test for notification name. See `TestWSExecutionVMStateCheck`.

File: pkg/services/rpcsrv/server.go (hunk 25 above)
Comment (AnnaShaleva) [reply pkg/services/rpcsrv/server.go:2767]: Create a separate test in `services/rpcsrv/subscription_test.go`, it may have a structure similar to `TestFilteredSubscriptions`, but we only need to check failed/successful response from `subscribe` RPC request.

**Review by AnnaShaleva — COMMENTED** [review]
Please, adjust the PR description, PR includes not only `int -> byte` filter field change.

File: pkg/neorpc/filters.go [hunk 27]
```diff
@@ -49,6 +54,16 @@ type (
...
 )
 
+// SubscriptionFilter is an interface for all subscription filters.
+type SubscriptionFilter interface {
+	// IsValid checks whether the filter is valid and returns
+	// a specific [ErrInvalidSubscriptionFilter] error if not.
+	IsValid() error
+}
+
+// ErrInvalidSubscriptionFilter is returned when the filter is invalid.
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:64]: s/the filter/the subscription filter

Just because we're at `neorpc` package, so from the user's POW it would be `neorpc.ErrInvalidSubscriptionFilter`, and documentation should be a little more precise.

File: pkg/rpcclient/wsclient_test.go [hunk 28]
```diff
@@ -373,6 +373,19 @@ func TestWSExecutionVMStateCheck(t *testing.T) {
 	wsc.Close()
 }
 
+func TestWSNotificationNameCheck(t *testing.T) {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/wsclient_test.go:376]: It's more likely `TestWSExecutionNotificationNameCheck`.

File: pkg/rpcclient/wsclient_test.go [hunk 29]
```diff
@@ -373,6 +373,19 @@ func TestWSExecutionVMStateCheck(t *testing.T) {
...
+	// Will answer successfully if request slips through.
+	srv := initTestServer(t, `{"jsonrpc": "2.0", "id": 1, "result": "55aaff00"}`)
+	wsc, err := NewWS(context.TODO(), httpURLtoWS(srv.URL), WSOptions{})
+	require.NoError(t, err)
+	wsc.getNextRequestID = getTestRequestID
+	require.NoError(t, wsc.Init())
+	filter := "notification_from_execution_with_long_name"
+	_, err = wsc.ReceiveExecutionNotifications(&neorpc.NotificationFilter{Name: &filter}, make(chan *state.ContainedNotificationEvent))
+	require.Error(t, err)
```
Comment (AnnaShaleva) [inline pkg/rpcclient/wsclient_test.go:385]: We now have dedicated error, thus, it can be checked directly. Try `require.ErrorIs(t, err, neorpc.ErrInvalidSubscriptionFilter)`, and for `TestWSExecutionVMStateCheck` as far.

File: pkg/services/rpcsrv/subscription_test.go [hunk 30]
```diff
@@ -625,3 +625,60 @@ func TestSubscriptionOverflow(t *testing.T) {
 	finishedFlag.CompareAndSwap(false, true)
 	c.Close()
 }
+
+func TestFilteredIsValidSubscriptions(t *testing.T) {
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/subscription_test.go:629]: `TestFilteredSubscriptions_InvalidFilter` (because that's exactly what you're testing) or `TestFilteredSubscriptions_FilterIsValid`?

File: pkg/services/rpcsrv/subscription_test.go [hunk 31]
```diff
@@ -625,3 +625,60 @@ func TestSubscriptionOverflow(t *testing.T) {
...
+	priv0 := testchain.PrivateKeyByID(0)
+	var goodSender = priv0.GetScriptHash()
+
+	var cases = map[string]struct {
+		params string
+	}{
+		"tx wrong sender": {
+			params: `["transaction_added", {"unknown_sender":"` + goodSender.StringLE() + `"}]`,
+		},
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/subscription_test.go:638]: This one is not related to the filter test cases, should be removed.

File: pkg/services/rpcsrv/subscription_test.go [hunk 32]
```diff
@@ -625,3 +625,60 @@ func TestSubscriptionOverflow(t *testing.T) {
...
+		"notification with long name": {
+			params: `["notification_from_execution", {"name":"notification_from_execution_with_long_name"}]`,
+		},
+		"execution not valid vm state": {
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/subscription_test.go:642]: `execution with invalid vm state`?

File: pkg/services/rpcsrv/subscription_test.go [hunk 33]
```diff
@@ -625,3 +625,60 @@ func TestSubscriptionOverflow(t *testing.T) {
...
+	}
+	var s string
+	for name, this := range cases {
+		t.Run(name, func(t *testing.T) {
+			chain, rpcSrv, c, respMsgs, finishedFlag := initCleanServerAndWSClient(t)
+
+			defer chain.Close()
+			defer rpcSrv.Shutdown()
+
+			blockSubID := callSubscribe(t, c, respMsgs, `["block_added"]`)
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/subscription_test.go:654]: You don't need block subscription for this test.

File: pkg/services/rpcsrv/subscription_test.go [hunk 34]
```diff
@@ -625,3 +625,60 @@ func TestSubscriptionOverflow(t *testing.T) {
...
+
+			resp := callWSGetRaw(t, c, fmt.Sprintf(`{"jsonrpc": "2.0","method": "subscribe","params": %s,"id": 1}`, this.params), respMsgs)
+			require.NotNil(t, resp.Error)
+			require.Nil(t, resp.Result)
+			require.Error(t, json.Unmarshal(resp.Result, &s))
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/subscription_test.go:659]: Don't need to check the result unmarshalling, result is nil, you've checked this above. Instead it's better check a specific `ErrInvalidSubscriptionFilter` error (using either errors.Is if possible or require.Contains for string error representation).

File: pkg/services/rpcsrv/subscription_test.go [hunk 35]
```diff
@@ -625,3 +625,60 @@ func TestSubscriptionOverflow(t *testing.T) {
...
+					index := rmap["index"].(float64)
+					if uint32(index) == lastBlock {
+						break
+					}
+					continue
+				}
+			}
+
+			callUnsubscribe(t, c, respMsgs, blockSubID)
+			finishedFlag.CompareAndSwap(false, true)
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/subscription_test.go:680]: You don't need this code, it's just unnecessary. The only thing you're checking here if failed "subscribe" request due to invalid filter.

File: pkg/services/rpcsrv/subscription_test.go [hunk 36]
```diff
@@ -625,3 +625,60 @@ func TestSubscriptionOverflow(t *testing.T) {
...
+			params: `["notification_from_execution", {"name":"notification_from_execution_with_long_name"}]`,
+		},
+		"execution not valid vm state": {
+			params: `["transaction_executed", {"state":"NOTHALT"}]`,
+		},
+	}
+	var s string
+	for name, this := range cases {
+		t.Run(name, func(t *testing.T) {
+			chain, rpcSrv, c, respMsgs, finishedFlag := initCleanServerAndWSClient(t)
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/subscription_test.go:649]: Move chain creation above a subsequent test run, because you don't need to create new chain for every test, it's enough to have a single instance. All our tests take too long to pass, so if possible, we should make new tests as fast as possible.

**Review by AnnaShaleva — COMMENTED** [review]
Otherwise LGTM.

File: docs/notifications.md [hunk 37]
```diff
@@ -64,7 +64,7 @@ omitted if empty).
 
 Recognized stream names:
  * `block_added`
-   Filter: `primary` as an integer with primary (speaker) node index from
+   Filter: `primary` as a byte with primary (speaker) node index from
```
Comment (roman-khimov) [inline docs/notifications.md:67]: It's JSON, it's an integer there.

File: pkg/neorpc/filters.go [hunk 38]
```diff
@@ -11,7 +11,7 @@ type (
 	// since/till the specified index inclusively). nil value treated as missing
 	// filter.
 	BlockFilter struct {
-		Primary *int    `json:"primary,omitempty"`
+		Primary *byte   `json:"primary,omitempty"`
```
Comment (roman-khimov) [inline pkg/neorpc/filters.go:19]: What happens on unmarshalling a JSON with `"primary": 100500`?

File: pkg/neorpc/filters.go [hunk 39]
```diff
@@ -49,6 +54,16 @@ type (
 	}
 )
 
+// SubscriptionFilter is an interface for all subscription filters.
+type SubscriptionFilter interface {
```
Comment (roman-khimov) [inline pkg/neorpc/filters.go:58]: Looks strange with the only validity check method. But maybe.

File: pkg/neorpc/filters.go (hunk 38 above)
Comment (AliceInHunterland) [reply pkg/neorpc/filters.go:19]: neorpc.Error{Code:-32602, Message:"Invalid params", Data:"json: cannot unmarshal number 100500 into Go struct field BlockFilter.primary of type uint8"

Comment (AnnaShaleva) [reply pkg/neorpc/filters.go:19]: I think it's OK, isn't it? That was the whole point of https://github.com/nspcc-dev/neo-go/issues/3241#issue-2015972066. @roman-khimov, don't you think so?

Comment (roman-khimov) [reply pkg/neorpc/filters.go:19]: OK, invalid is invalid.

File: docs/notifications.md [hunk 40]
```diff
@@ -85,7 +85,7 @@ Recognized stream names:
  * `notification_from_execution`
    Filter: `contract` field containing a string with hex-encoded Uint160 (LE
    representation) and/or `name` field containing a string with execution 
-   notification name.
+   notification name and should not be longer runtime.MaxEventNameLen value.
```
Comment (AnnaShaleva) [inline docs/notifications.md:88]: `and` doesn't suit here, use `which`.

Comment (AnnaShaleva) [inline docs/notifications.md:88]: not be longer than

File: pkg/rpcclient/wsclient.go [hunk 41]
```diff
@@ -127,7 +127,7 @@ func (r *blockReceiver) EventID() neorpc.EventID {
 }
 
 // Filter implements neorpc.Comparator interface.
-func (r *blockReceiver) Filter() any {
+func (r *blockReceiver) Filter() neorpc.SubscriptionFilter {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/wsclient.go:130]: Change the implementation accordingly. E.g. `func (r *headerOfAddedBlockReceiver) Filter() neorpc.SubscriptionFilter {` and ditto for the others implementations. Somehow I thought you had this code before, and now it's gone. The linter is unhappy about it and the project can't be built.

**Review by AnnaShaleva — COMMENTED** [review]
After all changes ensure that the Linter job passes.

File: pkg/services/rpcsrv/subscription_test.go [hunk 42]
```diff
@@ -676,3 +676,31 @@ func TestSubscriptionOverflow(t *testing.T) {
...
+	chain, rpcSrv, c, respMsgs, finishedFlag := initCleanServerAndWSClient(t)
+	defer chain.Close()
+	defer rpcSrv.Shutdown()
+
+	for name, this := range cases {
+		t.Run(name, func(t *testing.T) {
+			resp := callWSGetRaw(t, c, fmt.Sprintf(`{"jsonrpc": "2.0","method": "subscribe","params": %s,"id": 1}`, this.params), respMsgs)
+			require.NotNil(t, resp.Error)
+			require.Nil(t, resp.Result)
+			fmt.Println(neorpc.ErrInvalidSubscriptionFilter.Error())
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/subscription_test.go:700]: Remove `fmt.Println`.

Comment (AliceInHunterland) [reply pkg/services/rpcsrv/subscription_test.go]: fixed

File: docs/notifications.md [hunk 43]
```diff
@@ -85,7 +85,8 @@ Recognized stream names:
  * `notification_from_execution`
    Filter: `contract` field containing a string with hex-encoded Uint160 (LE
    representation) and/or `name` field containing a string with execution 
-   notification name.
+   notification name which should not be longer than runtime.MaxEventNameLen
```
Comment (roman-khimov) [inline docs/notifications.md:88]: The statement is NeoGo-specific, but we want this document to be generic, so we better have some more generic reference. Specific value can be added as well, these constants rarely change.

File: docs/notifications.md [hunk 44]
```diff
@@ -67,11 +67,11 @@ omitted if empty).
 
 Recognized stream names:
  * `block_added`
-   Filter: `primary` as an integer with primary (speaker) node index from
-   ConsensusData and/or `since` field as an integer value with block
-   index starting from which new block notifications will be received and/or
-   `till` field as an integer values containing block index till which new
-   block notifications will be received.
+   Filter: `primary` as an integer which will be converted to byte with
```
Comment (roman-khimov) [inline docs/notifications.md:70]: Just an integer wih a valid range of 0-255.

File: docs/notifications.md [hunk 45]
```diff
@@ -85,7 +85,7 @@ Recognized stream names:
  * `notification_from_execution`
    Filter: `contract` field containing a string with hex-encoded Uint160 (LE
    representation) and/or `name` field containing a string with execution 
-   notification name.
+   notification name which should not be longer than 32 symbols.
```
Comment (AnnaShaleva) [inline docs/notifications.md:88]: Symbol is not equal to byte, read the https://go.dev/blog/strings:
```
This technique is handy when much of the string is intelligible as text but there are peculiarities to root out; it produces:

"\xbd\xb2=\xbc ⌘"

If we squint at that, we can see that buried in the noise is one ASCII equals sign, along with a regular space, and at the end appears the well-known Swedish “Place of Interest” symbol. That symbol has Unicode value U+2318, encoded as UTF-8 by the bytes after the space (hex value 20): e2 8c 98.
```
Should be something like "which should be a valid UTF-8 string not longer than 32 bytes"

**Review by AnnaShaleva — COMMENTED** [review]
Otherwise LGTM.

File: pkg/neorpc/filters.go (hunk 39 above)
Comment (AnnaShaleva) [reply pkg/neorpc/filters.go:58]: We may extend it in future with the rest of filter-related methods, but it requires refactoring of the whole set of these methods.

File: docs/notifications.md (hunk 45 above)
Comment (AliceInHunterland) [reply docs/notifications.md]: fixed

File: pkg/neorpc/filters.go [hunk 46]
```diff
@@ -49,6 +54,16 @@ type (
...
+	// IsValid checks whether the filter is valid and returns
+	// a specific [ErrInvalidSubscriptionFilter] error if not.
+	IsValid() error
+}
+
+// neorpc.ErrInvalidSubscriptionFilter is returned when the subscription filter is invalid.
+var ErrInvalidSubscriptionFilter = errors.New("invalid subscription filter")
```
Comment (AnnaShaleva) [inline pkg/neorpc/filters.go:65]: Exported comment should start with `ErrInvalidSubscriptionFilter`, you already in the `neorpc` package.

Comment (AnnaShaleva) [reply pkg/neorpc/filters.go:65]: @AliceInHunterland, fix this, please.

Comment (AliceInHunterland) [reply pkg/neorpc/filters.go:65]: https://github.com/nspcc-dev/neo-go/pull/3258#discussion_r1432521382 What does than little more specific mean?

Comment (AliceInHunterland) [reply pkg/neorpc/filters.go]: fixed

**Review by AnnaShaleva — APPROVED** [review]
